File: 2025/05 May/23 Dice throw/main.py
# Link: https://www.geeksforgeeks.org/problems/dice-throw5349/1


class Solution:
    def noOfWays(self, m, n, x):
        dp = [0] * (x + 1)
        dp[0] = 1
        for i in range(1, n + 1):
            for j in range(x, 0, -1):
                dp[j] = 0
                for k in range(1, m + 1):
                    if j - k >= 0:
                        dp[j] += dp[j - k]

            dp[0] = 0
        return dp[x]


if __name__ == "__main__":
    obj = Solution()
    m = 6
    n = 3
    x = 12
    print(obj.noOfWays(m, n, x))

    m = 2
    n = 3
    x = 6
    print(obj.noOfWays(m, n, x))

File: 2025/05 May/23 Dice throw/main.cpp
// Link: https://www.geeksforgeeks.org/problems/dice-throw5349/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int noOfWays(int m, int n, int x)
    {
        vector<int> dp(x + 1, 0);

        dp[0] = 1;
        for (int i = 1; i <= n; i++)
        {
            for (int j = x; j >= 1; j--)
            {
                dp[j] = 0;
                for (int k = 1; k <= m && j - k >= 0; k++)
                {
                    dp[j] += dp[j - k];
                }
            }
            dp[0] = 0;
        }
        return dp[x];
    }
};

int main()
{
    Solution sol;
    int m = 6;
    int n = 3;
    int x = 12;
    cout << "Number of ways to get sum " << x << " with " << n << " dice: " << sol.noOfWays(m, n, x) << endl;

    m = 2;
    n = 3;
    x = 6;
    cout << "Number of ways to get sum " << x << " with " << n << " dice: " << sol.noOfWays(m, n, x) << endl;
    return 0;
}